iwyu.cc

```diff
@@ -1585,10 +1598,23 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       return;
 
     // Figure out the best location to attribute uses inside macros.
-    if (IsInMacro(used_loc))
-      used_loc = GetUseLocationForMacroExpansion(used_loc, decl);
     const FileEntry* used_in = GetFileEntry(used_loc);
+    if (IsInMacro(used_loc)) {
+      const NamedDecl* fwd_decl = GetForwardDeclareInSameFile(decl, used_loc);
+      if (fwd_decl) {
+        // Make sure we keep the heuristic forward decl.
+        preprocessor_info().FileInfoFor(used_in)->ReportForwardDeclareUse(
+            used_loc, decl, IsNodeInsideCXXMethodBody(current_ast_node()),
+            comment);
+
+        // Attribute use to caller
+        used_loc = GetInstantiationLoc(used_loc);
+      } else {
+        used_loc = GetUseLocationForMacroExpansion(used_loc, decl);
+      }
+    }
 
+    used_in = GetFileEntry(used_loc);
```

Comment:
You could set `used_in` in different ways, for example, inside previous if-block. But it doesn't matter, it's just a matter of taste and I have no arguments that some of the approaches is better, the only difference so far is probably nicer diff.